dm/syncer/validator_checkpoint.go

```diff
@@ -0,0 +1,436 @@
+			binlog_pos INT UNSIGNED,
+			binlog_gtid TEXT,
+			create_time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+			update_time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+			UNIQUE KEY uk_source (source)
+		)`,
+		`CREATE TABLE IF NOT EXISTS ` + c.pendingChangeTableName + ` (
+			id BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,
+			source VARCHAR(32) NOT NULL,
+			schema_name VARCHAR(128) NOT NULL,
+			table_name VARCHAR(128) NOT NULL,
+			row_pk VARCHAR(128) NOT NULL,
+			data JSON NOT NULL,
+			revision bigint NOT NULL,
+			create_time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+			update_time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+			UNIQUE KEY uk_source_schema_table_key(source, schema_name, table_name, row_pk),
+			INDEX idx_revision(revision)
+		)`,
+		`CREATE TABLE IF NOT EXISTS ` + c.errorChangeTableName + ` (
+			id BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,
+			source VARCHAR(32) NOT NULL,
+			src_schema_name VARCHAR(128) NOT NULL,
+			src_table_name VARCHAR(128) NOT NULL,
+			row_pk VARCHAR(128) NOT NULL,
+			dst_schema_name VARCHAR(128) NOT NULL,
+			dst_table_name VARCHAR(128) NOT NULL,
+			data JSON NOT NULL,
+			dst_data JSON NOT NULL,
+			error_type int NOT NULL,
+			status int not null,
+			create_time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+			update_time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+			UNIQUE KEY uk_source_schema_table_key(source, src_schema_name, src_table_name, row_pk)
+		)`,
+		`CREATE TABLE IF NOT EXISTS ` + c.tableStatusTableName + ` (
+			id BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,
+			source VARCHAR(32) NOT NULL,
+			src_schema_name VARCHAR(128) NOT NULL,
+			src_table_name VARCHAR(128) NOT NULL,
+			dst_schema_name VARCHAR(128) NOT NULL,
+			dst_table_name VARCHAR(128) NOT NULL,
+			stage VARCHAR(32) NOT NULL,
+            message VARCHAR(512) NOT NULL,
+			create_time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+			update_time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+			UNIQUE KEY uk_source_schema_table_key(source, src_schema_name, src_table_name)
+		)`,
+	}
+	_, err := c.dbConn.ExecuteSQL(tctx, sqls)
+	tctx.L().Info("create checkpoint and data table", zap.Strings("statements", sqls))
+	return err
+}
+
+func (c *validatorPersistHelper) persist(loc binlog.Location) error {
+	count := len(c.validator.tableStatus)
+	for _, worker := range c.validator.workers {
+		for _, tblChange := range worker.getPendingChangesMap() {
+			count += len(tblChange.rows)
+		}
+		count += len(worker.errorRows)
+	}
+	queries := make([]string, 0, count+2)
+	args := make([][]interface{}, 0, count+2)
+	nextRevision := c.revision + 1
+
+	c.tctx.L().Info("persist checkpoint and intermediate data")
+
+	// update checkpoint
+	queries = append(queries, `INSERT INTO `+c.checkpointTableName+
+		`(source, binlog_name, binlog_pos, binlog_gtid) VALUES (?, ?, ?, ?) ON DUPLICATE KEY UPDATE
```

Comment:
considering that the amount of data here is much larger than the other checkpoints 
i am a little concerned about performance issues about always use `Insert ON DUPLICATE KEY UPDATE` 

see https://asktug.com/t/topic/33147

Reply:
add a todo